buffer.go

```diff
@@ -306,45 +307,47 @@ func (b *buffer) unref() {
 	}
 }
 
-func (b *buffer) reset() {
-	b.data = b.data[:0]
-}
-
-func (b *buffer) resize(size int) {
-	if cap(b.data) < size {
-		const pageSize = 4096
-		minSize := 2 * cap(b.data)
-		bufferSize := ((size + (pageSize - 1)) / pageSize) * pageSize
-		if bufferSize < minSize {
-			bufferSize = minSize
-		}
-		b.data = make([]byte, size, bufferSize)
-	} else {
-		b.data = b.data[:size]
-	}
-}
-
-func (b *buffer) clone() (clone *buffer) {
-	if b.pool != nil {
-		clone = b.pool.get()
-	} else {
-		clone = &buffer{refc: 1}
-	}
-	clone.data = append(clone.data, b.data...)
-	return clone
-}
+// bufferPool holds a slice of sync.pools used for levelled buffering.
+// the table below shows the pools used for different buffer sizes when both getting
+// and putting a buffer. when allocating a new buffer from a given pool we always choose the
+// min of the put range to guarantee that all gets will have an adequately sized buffer.
+//
+// [pool] : <get range>  : <put range>  : <alloc size>
+// [0]    : 0    -> 1023 : 1024 -> 2047 : 1024
+// [1]    : 1024 -> 2047 : 2048 -> 4095 : 2048
+// [2]    : 2048 -> 4095 : 4096 -> 8191 : 4096
+// ...
+const numPoolBuckets = 16
+const basePoolIncrement = 1024
 
 type bufferPool struct {
-	pool sync.Pool
+	pool [numPoolBuckets]sync.Pool
 }
 
-func (p *bufferPool) get() *buffer {
-	b, _ := p.pool.Get().(*buffer)
+// get returns a buffer from the levelled buffer pool. sz is used to choose the appropriate pool
+func (p *bufferPool) get(sz int) *buffer {
+	i := levelledPoolIndex(sz)
+	b, _ := p.pool[i].Get().(*buffer)
 	if b == nil {
-		b = &buffer{pool: p}
-	} else {
-		b.reset()
+		// align size to the pool
+		poolSize := basePoolIncrement << i
+		if sz > poolSize { // this can occur when the buffer requested is larger than the largest pool
+			poolSize = sz
+		}
+		b = &buffer{
+			data: make([]byte, 0, poolSize),
+			pool: p,
+		}
 	}
+	// if the buffer comes from the largest pool it may not be big enough
+	if cap(b.data) < sz {
+		p.pool[i].Put(b)
+		b = &buffer{
+			data: make([]byte, 0, sz),
```

Comment:
Should we align `sz` on the bucket upper bound here?

Reply:
technically this shouldn't be necessary. the only time that `cap(b.data) < sz` could be true is if sz > 32MB and the pool happens to pull a buffer that isn't large enough. however, since this behavior is spread across two functions and is far from obvious i have no issues adding code directly here to support that if you'd prefer.

Reply:
OK, let's try it out as-is and see if we ever measure issues with this approach 👍